[feat] 사용자별 추천 기록 생성 API 구현

zerone1202 opened the pull request:
## #️⃣ Issue Number<!--- ex) #이슈번호, #이슈번호 --> 

- Closes #222
- Recommendation 생성 API를 구현했습니다. 
- 모든 옷 데이터를 동기적으로 조회하여 1년 이상 미착용된 옷에 대한 추천 기록을 생성하며, 이는 향후 성능 고도화를 위한 기준점 역할을 합니다.

## 📝 요약(Summary)<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
- Service 인터페이스 확립: RecommendationCommandService 인터페이스와 구현체를 정의하여 추천 생성 로직을 캡슐화
- 핵심 로직 구현: 1년 이상 미착용된 옷에 대해 판매(SALE}와 기부(DONATION) 두 종류의 추천 기록을 생성하고 saveAll()로 일괄 저장

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)

- 성공 응답 (추천 대상이 없어서 빈 리스트 반환)
<img width="1408" height="614" alt="스크린샷 2025-11-02 03 39 18" src="https://github.com/user-attachments/assets/f91bbdfb-cbaa-47a1-a5ac-900f08391df0" />

- 에러 응답
<img width="1408" height="510" alt="스크린샷 2025-11-02 03 40 43" src="https://github.com/user-attachments/assets/b6e361db-572b-44ff-8eac-818ce4688b79" />

## 💬 공유사항 to 리뷰어<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->
- 동기식 로직: 현재 generateRecommendations 로직은 향후 배치 고도화를 위한 성능 기준선입니다. for-loop 대신 Stream을 사용했지만, 대량 데이터에 대한 성능 부하는 Scheduler 및 Batch에서 개선될 예정입니다.

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [X] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

zerone1202 committed:
[feat] recommendation: 도메인 예외 처리 클래스 RecommendationException 추가

- RecommendationErrorCode를 사용하여 도메인별 예외를 처리하는 커스텀 Exception 클래스 `RecommendationException`을 정의
- 이를 통해 도메인 비즈니스 로직 내에서 발생하는 오류를 일관된 형식으로 처리하도록 구조화

zerone1202 committed:
[feat] recommendation: 도메인 오류 코드 RecommendationErrorCode 정의

- ErrorCode 인터페이스를 구현하여 Recommendation 도메인에서 발생하는 모든 오류를 관리하는 Enum 클래스 `RecommendationErrorCode` 추가
- 이를 통해 오류 코드(code), HTTP 상태(httpStatus), 메시지(message)를 일관성 있게 중앙 관리하도록 구조화

zerone1202 committed:
[feat] recommendation: 도메인 성공 코드 RecommendationSuccessCode 정의

- SuccessCode 인터페이스를 구현하여 Recommendation 도메인에서 발생하는 모든 성공을 관리하는 Enum 클래스 `RecommendationSuccessCode` 추가
- 이를 통해 성공 코드(code), HTTP 상태(httpStatus), 메시지(message)를 일관성 있게 중앙 관리하도록 구조화

zerone1202 committed:
[feat] recommendation: 추천 기록 생성 응답 DTO RecommendationCreateResponse 정의

- Recommendation 엔티티를 기반으로 생성 후 응답에 필요한 모든 필드를 포함하는 Record 타입 DTO 정의
- from(Recommendation recommendation) 정적 팩토리 메서드를 구현하여 엔티티에서 DTO로의 변환 로직을 캡슐화

zerone1202 committed:
[feat] recommendation: RecommendationCommandService 인터페이스 정의

- 추천 생성 메서드 `generateRecommendations(Long userId)`를 정의
- 반환 타입을 최종 응답 DTO 목록인 `List<RecommendationCreateResponse>`으로 확정하여 Service 계층의 인터페이스를 명확히 함

zerone1202 committed:
[feat] recommendation: RecommendationCommandServiceImpl 구현 완료 (동기식 기준선)

- RecommendationCommandService의 핵심 메서드 `generateRecommendations(Long userId)` 구현
- UserQueryService와 ClothesQueryService를 사용하여 도메인 경계를 준수하며 데이터 조회
- 1년 이상 미착용 옷을 필터링하고, 해당 옷에 대해 판매(SALE)와 기부(DONATION) 두 종류의 추천 기록을 일괄 생성하도록 로직 구현
- 최종적으로 생성된 엔티티를 RecommendationCreateResponse DTO 목록으로 변환하여 반환

zerone1202 committed:
[feat] recommendation: RecommendationController에 추천 기록 수동 생성 API 추가

- POST /v1/recommendations 엔드포인트를 통해 RecommendationCommandService를 호출하는 API 구현
- @AuthenticationPrincipal을 사용하여 인증된 사용자 ID 기반으로 추천 기록을 동기적으로 생성
- Swagger Operation 및 ApiResponse를 명확히 정의하여 API 명세를 완성하고, RecommendationSuccessCode를 사용하여 성공 응답을 반환
- 이 API는 Scheduler 도입 전, 성능 기준선 확보 및 수동 테스트용으로 사용됨

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/command/RecommendationCommandServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+        // lastWornAt을 LocalDate로 변환하여 1년 전과 비교
+        LocalDate lastWornDate = lastWornAt.toLocalDate();
+        LocalDate oneYearAgo = LocalDate.now().minusYears(1);
```
적용하겠습니다.

zerone1202 replied:
시스템 기본 시간대에 의존하지 않고 주입된 Clock을 사용하여 현재 시간을 가져오는 걸로 수정하였습니다!

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/command/RecommendationCommandServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+                                clothes,
+                                RecommendationType.SALE,
+                                "마지막 착용일이 1년 이상 경과"
+                        ),
+                        // 기부 추천
+                        Recommendation.createForUnwornClothes(
+                                user,
+                                clothes,
+                                RecommendationType.DONATION,
+                                "마지막 착용일이 1년 이상 경과"
+                        )
+                ).stream())
+                .toList();
```
모든 추천 타입을 순회하도록 변경하고,
추천 사유 문자열을 private static final 상수로 추출하는 것으로 수정하겠습니다.

AllSungho reviewed (approved):
늦은 시각에도 고생이 많습니다!!
수고하셨습니다!!!

zerone1202 committed:
[feat] recommendation: 추천 생성 로직 안정성 및 확장성 개선

- Clock을 주입받아 LocalDate 비교에 사용하도록 변경하여 시간 의존성 제거 및 테스트 용이성 확보
- RecommendationType.values()를 순회하여 동적으로 추천 기록을 생성하도록 변경, 향후 추천 타입 확장 시 Service 수정 불필요
- 추천 사유 문자열을 private static final 상수로 추출하여 코드 가독성 및 유지보수성 향상

AllSungho reviewed (approved):
수정사항 확인했습니다!

zerone1202 committed:
[feat] recommendation: 테스트 용이성을 위한 Clock Bean 정의 추가

- common.config 패키지에 ClockConfig 클래스를 추가하여 java.time.Clock Bean을 등록
- Clock.systemDefaultZone()을 기본 Clock으로 제공하여 런타임에서 시간을 주입하고, 테스트 시 Mocking을 용이하게 함으로써 코드 안정성을 확보

AllSungho reviewed (approved):
CI 확인했습니다!
